rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1614 @@
+    if cond1 then
+      foo
+    else if cond2 then
+      bar
+    else
+      baz
+    ```
+
+- Indentation should reflect the expression structure.
+  Example:
+  ```nix
+  # Bad, the indentation misleads the user
+  {
+    foo = {
+    bar = if
+    baz == null then 10
+      else 20
+    ;
+  }; }
+
+  # Good
+  {
+    foo = {
+      bar =
+        if baz == null then
+          10
+        else
+          20;
+    };
+  }
+  ```
+
+### Editor Config
+
+This [editor config](https://editorconfig.org/) specifies the basic details about Nix files:
+
+```editorconfig
+end_of_line = lf
+insert_final_newline = true
+trim_trailing_whitespace = true
+charset = utf-8
+indent_style = space
+```
+
+### Single-line common ancestor expression rule
+
+For any two (sub-)expressions that are fully on a common single line, their smallest common ancestor expression must also be on the same line.
+
+**Example**
+
+```nix
+# Bad, expressions cond and foo are fully on the same line,
+# but their smallest common ancestor expression is the entire if-then-else, which spans multiple lines
+if cond then foo
+else bar
+
+# Okay, cond, foo and bar have the if-then-else as a common ancestor expression,
+# which is also fully on the same line
+if cond then foo else bar
+
+# Bad, due to function application precedence, the smallest common ancestor expression
+# of foo and bar is `foo || bar baz`, which spans two lines
+foo || bar
+  baz
+```
+
+**Rationale**
+
+This rule has turned out to be very practical at catching code that could be potentially hard to understand or edit.
+
+### Line length
+
+- There should be a configurable _soft_ line length limit, limiting the number of characters on one line without counting the leading indentation.
+  The default should be 100 characters.
+- There may also be a configurable _hard_ line length limit, which includes the leading indentation.
+- String-like values such as strings, paths, comments, urls, etc. may go over the hard line length limit.
+
+### Indentation
+
+- Two spaces must be used for each indentation level.
+  - This may be revisited should Nix get proper support for [using tabs for indentation](https://github.com/NixOS/nix/issues/7834) in the future.
+- Vertical alignment must be ignored, both at the start of the line and within lines.
+  - Examples:
+    ```nix
+    {
+      # Bad, vertical alignment Within lines
+      linux    = { execFormat = elf;     families = {              }; };
+      netbsd   = { execFormat = elf;     families = { inherit bsd; }; };
+      none     = { execFormat = unknown; families = {              }; };
+      openbsd  = { execFormat = elf;     families = { inherit bsd; }; };
+
+      # Bad, vertical alignment at the start of line
+      optExecFormat =
+        lib.optionalString (kernel.name == "netbsd" &&
+                            gnuNetBSDDefaultExecFormat cpu != kernel.execFormat
+                           )
+                           kernel.execFormat.name;
+    }
```

Comment:
For sections like these, it'd be great if you could make explicit exceptions to this rule. There's like a handful of places this kind of tabular style really has its benefits, most of them in the systems declarations.

Reply:
Nix language imho plays the declarative config management role here. Other declarative config DSLs allow (and some even require) vertical "tabular" vertical alignment as in the `linux, netbsd, ...` case, examples being Puppet or OpenTofu (Terraform) where the latter is even conceptually quite close to Nix.

It is unclear to me whether this kind of alignment is "bad" or "discouraged" and/or whether the formatter should or should not try to "fix" it.

The statement says that formatter should in fact _ignore_ this kind of vertical alignment which to me means that after the code is formatted, vertical alignment should be _preserved as is_. On the other hand the example code says vertical alignment is "bad" and so it is (to me at least) ultimately unclear what the point is.

I personally like and use vertical alignment as in the first example because it (imho) _significantly_ increases human readability of the _declarative config_ code by allowing much faster and effective "eye scanning" of the code.

Note that I'm talking about the _config management_ role of the Nix language, not the _programming_ role - it is imho important to remember that Nix has multiple roles which may justify different formatting styles based on the context the code is used in.

If the statement (as I hope) really means that vertical alignment should be preserved by the formatter then I suggest removing the comment stating that the formatting is "bad".

If not then I would very much like to see some objective reasoning as to _why_ is it "bad", because to me, obviously, it is actually (very) good.

Reply:
> The statement says that formatter should in fact ignore this kind of vertical alignment which to me means that after the code is formatted, vertical alignment should be preserved as is.

I agree that the wording of the text needs to be adjusted, because this is certainly not its intent.

Let's put the discussion about whether we want vertically aligned code or not aside, and instead focus on on the *how*. Because simply saying "I want this specific code preserved" is not a rule that can be implemented. Everything needs to be algorithmically specified, and there is a strong requirement that doing this should not reduce the format quality for unaffected code. This is not a trivial problem.

We would probably also need a proper specification of what vertical alignment actually should look like, because if you look at Nix code you'll see various different styles being used. Since the point of this RFC is to unify the code style, this would have to be handled as well.

So even if we added an exception to allow for handling vertical alignment in some way (as we recently have done for selected other scenarios), this would still need someone to put the effort of actually implementing it.

Reply:
> simply saying "I want this specific code preserved" is not a rule that can be implemented.

fair point

> a proper specification of what vertical alignment actually should look like, 

that's where I think the different contexts of where Nix lang is used comes in (_declarative config_ vs _programming_ vs possibly other use cases I can't think of right now) and I agree it is far from trivial, if not impossible to infer the context from the code alone without some kind of "tagging" or "metas" or somesuch. And that would be bad.

> Let's ... focus on on the how. 

I have a very naive perspective on this but one of the options I see is to, at least at first, _not try_ to solve the problem of _how should it look like_ and rather just trust the code owner to do the inline vertical alignment properly according to whatever the use case of the code is. The algorithmic part then would be rather simple one: formatter should ignore any whitespaces between symbols within a line, except at the beginning and the end of the line. This could even be an optional rule in the formatter that must be explicitly enabled by a flag or config with the intention to keep the rules clear for any upstream which chooses to adhere to them while still allowing private use cases to leverage the formatter without breaking their style.

I believe that this might not be so hard to implement if the formatter code is designed around individual rules.

Reply:
We now updated this part with better justification and more examples: https://github.com/NixOS/rfcs/pull/166/commits/26b444b58db94baa2636d3b940a90768745e8692

Reply:
One rather simple practical method of allowing free-form formatting for certain sections of code that just sprung to my mind would be to do so on file-level granularity. A certain comment text in the first line would simply disable the auto-formatter for the entire file. If the formatter would encounter such a comment, the entire file would be considered formatted correctly and remain untouched.

Sections like the systems example could then simply be placed into a dedicated file, formatted with vertical alignment and whatever other non-conformant arbitrary rules without upsetting future CI.

I imagine this could cover most of the exceptional cases where vertical alignment is absolutely necessary.